realm/realm-library/src/main/cpp/io_realm_internal_OsSet.cpp

```diff
@@ -0,0 +1,139 @@
+
+void finalize_set(jlong ptr) {
+    delete reinterpret_cast<object_store::Set*>(ptr);
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsSet_nativeGetFinalizerPtr(JNIEnv*, jclass) {
+    return reinterpret_cast<jlong>(&finalize_set);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_OsSet_nativeCreate(JNIEnv* env, jclass, jlong shared_realm_ptr,
+                                          jlong obj_ptr, jlong column_key) {
+    try {
+        auto obj = *reinterpret_cast<realm::Obj*>(obj_ptr);
+        auto shared_realm = *reinterpret_cast<SharedRealm*>(shared_realm_ptr);
+        auto set_ptr = new object_store::Set(shared_realm, obj, ColKey(column_key));
+        return reinterpret_cast<jlong>(set_ptr);
+    }
+    CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_OsSet_nativeSize(JNIEnv* env, jclass, jlong set_ptr) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        return set.size();
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_OsSet_nativeContainsNull(JNIEnv *env, jclass, jlong set_ptr) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        size_t found = set.find_any(Mixed());
+        return found;
+    }
+    CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_OsSet_nativeContainsString(JNIEnv* env, jclass, jlong set_ptr,
+                                                  jstring j_value) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        JStringAccessor value(env, j_value);
+        size_t found = set.find_any(Mixed(StringData(value)));
+        return found;
+    }
+    CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_OsSet_nativeAddNull(JNIEnv* env, jclass, jlong set_ptr) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        set.insert(Mixed());
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_OsSet_nativeAddString(JNIEnv* env, jclass, jlong set_ptr, jstring j_value) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        JStringAccessor value(env, j_value);
+        JavaAccessorContext context(env);
+        set.insert(context, Any(value));
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT jlongArray JNICALL
+Java_io_realm_internal_OsSet_nativeRemoveNull(JNIEnv* env, jclass, jlong set_ptr) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        const std::pair<size_t, bool>& remove_pair = set.remove_any(Mixed());
+        jlong ret[2];
+        ret[0] = remove_pair.first;     // index
+        ret[1] = remove_pair.second;    // found (or not)
+        jlongArray ret_array = env->NewLongArray(2);
```

Comment:
Would probably have used a special value (`-1`) to indicate not found. But since core is returning it like that there are maybe odd semantics that cannot be represented that way 🤷

Reply:
I'm a bit concerned about these constants. There is also the `npos` value that is returned if an item is "not found", which has no representation in the SDK... I think we will just have to tread lightly with this.